llmfoundry/tokenizers/tiktoken.py

```diff
@@ -0,0 +1,290 @@
+# Copyright 2022 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import torch
+from transformers import PreTrainedTokenizer
+
+
+class TiktokenTokenizerWrapper(PreTrainedTokenizer):
+    """A thin wrapper around tiktoken to make it compatible with Hugging Face.
+
+    tokenizers.
+
+    See HuggingFace for further documentation on general tokenizer methods.
+    """
+
+    model_input_names = ['input_ids', 'attention_mask']
+
+    def __init__(self,
+                 model_name: Optional[str] = None,
+                 encoding_name: Optional[str] = None,
+                 add_bos_token: bool = False,
+                 unk_token: Optional[str] = '<|endoftext|>',
+                 eos_token: Optional[str] = '<|endoftext|>',
+                 bos_token: Optional[str] = '<|endoftext|>',
+                 pad_token: Optional[str] = None,
+                 **kwargs: Dict[str, Any]):
+        """Constructor creates a tiktoken tokenizer to use as the underlying.
+
+        tokenizer.
+
+        Args:
+            model_name (Optional[str], optional): The name of the model to load from tiktoken. Defaults to None.
+                Either model_name or encoding_name must be set, but not both.
+            encoding_name (Optional[str], optional): The name of the encoding to load from tiktoken. Defaults to None.
+                Either model_name or encoding_name must be set, but not both.
+            add_bos_token (bool, optional): Whether to add bos tokens. Defaults to False.
+            unk_token (Optional[str], optional): The unk token. Defaults to '<|endoftext|>'.
+            eos_token (Optional[str], optional): The eos token. Defaults to '<|endoftext|>'.
+            bos_token (Optional[str], optional): The bos token. Defaults to '<|endoftext|>'.
+            pad_token (Optional[str], optional): The pad token. Defaults to None.
+        """
+        try:
+            import tiktoken
+        except:
+            raise ImportError(
+                'You need to install tiktoken to use TiktokenTokenizerWrapper.')
+
+        if model_name is not None and encoding_name is not None:
+            raise ValueError(
+                'You need to specify either model_name or encoding_name, not both.'
+            )
```

Comment:
Would it make sense to instead set a `encoding_name` default if both are None?

Reply:
mm, i'd rather force the user to consciously pick a tokenizer.